File: engines/songwriting_engine/runtime/musicxml_exporter.py
"""
MusicXML Exporter — Lead sheet with melody, chords, lyrics, section markers, melisma.
"""

from typing import Any, Dict, List, Optional

DIVISIONS = 4
ENHARMONIC = {
    0: ("C", 0), 1: ("C", 1), 2: ("D", 0), 3: ("D", 1), 4: ("E", 0), 5: ("F", 0),
    6: ("F", 1), 7: ("G", 0), 8: ("G", 1), 9: ("A", 0), 10: ("A", 1), 11: ("B", 0),
}


def _duration_to_xml(duration: float) -> tuple:
    divs = max(1, int(duration * DIVISIONS))
    if divs <= 1:
        return 1, "16th", False
    if divs <= 2:
        return 2, "eighth", False
    if divs <= 4:
        return 4, "quarter", False
    if divs <= 6:
        return 6, "quarter", True
    if divs <= 8:
        return 8, "half", False
    if divs <= 12:
        return 12, "half", True
    return 16, "whole", False


def _spell_pitch(midi: int) -> tuple:
    pc = midi % 12
    octave = (midi // 12) - 1
    step, alter = ENHARMONIC.get(pc, ("C", 0))
    return step, alter, octave


def _note_xml(pitch: int, duration: float, syllable: Optional[str] = None, syllabic: Optional[str] = None) -> str:
    divs, typ, dot = _duration_to_xml(duration)
    step, alter, octave = _spell_pitch(pitch)
    alter_tag = f"<alter>{alter}</alter>" if alter != 0 else ""
    dot_tag = "<dot/>" if dot else ""
    lyric = ""
    if syllable:
        syllabic_val = syllabic or "single"
        lyric = f"<lyric><syllabic>{syllabic_val}</syllabic><text>{syllable}</text></lyric>"
    elif syllabic in ("middle", "end"):
        lyric = f"<lyric><syllabic>{syllabic}</syllabic></lyric>"
    return (
        f"      <note><pitch><step>{step}</step>{alter_tag}<octave>{octave}</octave></pitch>"
        f"<duration>{divs}</duration><type>{typ}</type>{dot_tag}"
        f"{lyric}</note>"
    )


def _rest_xml(duration: float) -> str:
    divs, typ, dot = _duration_to_xml(duration)
    dot_tag = "<dot/>" if dot else ""
    return f"      <note><rest/><duration>{divs}</duration><type>{typ}</type>{dot_tag}</note>"


class MusicXMLExporter:
    """Export CandidateComposition to MusicXML lead sheet."""

    def export(
        self,
        candidate: Dict[str, Any],
        title: Optional[str] = None,
        key_center: str = "C",
    ) -> str:
        """Produce MusicXML string."""
        title = title or candidate.get("title", "Song")
        sections = candidate.get("sections", [])
        melody = candidate.get("melody", [])
        harmony = candidate.get("harmony", [])

        by_measure: Dict[int, List[Dict]] = {}
        for e in melody:
            m = e.get("measure", 0)
            if m not in by_measure:
                by_measure[m] = []
            by_measure[m].append(e)

        chords_by_measure: Dict[int, List[Dict]] = {}
        for h in harmony:
            m = h.get("measure", 0)
            if m not in chords_by_measure:
                chords_by_measure[m] = []
            chords_by_measure[m].append(h)

        section_markers = {}
        for s in sections:
            bar_start = s.get("bar_start", 0)
            sid = s.get("id", "verse")
            section_markers[bar_start] = sid

        total_bars = max(by_measure.keys(), default=0) + 1
        total_bars = max(total_bars, max(chords_by_measure.keys(), default=0) + 1)

        lines = [
            '<?xml version="1.0" encoding="UTF-8"?>',
            '<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 3.1 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">',
            '<score-partwise version="3.1">',
            '  <work><work-title>' + title + '</work-title></work>',
            '  <identification><encoding><software>Songwriting Engine</software></encoding></identification>',
            '  <part-list>',
            '    <score-part id="P1"><part-name>Voice</part-name></score-part>',
            '  </part-list>',
            '  <part id="P1">',
        ]

        for m in range(total_bars):
            if m in section_markers:
                lines.append(f'    <measure number="{m + 1}">')
                lines.append(f'      <direction><direction-type><rehearsal>{section_markers[m]}</rehearsal></direction-type></direction>')
            else:
                lines.append(f'    <measure number="{m + 1}">')

            if m == 0:
                lines.append('      <attributes><divisions>4</divisions><key><fifths>0</fifths></key><time><beats>4</beats><beat-type>4</beat-type></time><clef><sign>G</sign><line>2</line></clef></attributes>')

            # Chord symbols
            for h in chords_by_measure.get(m, []):
                sym = h.get("symbol", "C")
                lines.append(f'      <harmony><root><root-step>{sym[0] if sym else "C"}</root-step></root></harmony>')

            # Notes
            evs = sorted(by_measure.get(m, []), key=lambda x: x.get("beat_position", 0))
            cursor = 0.0
            for e in evs:
                onset = e.get("beat_position", 0)
                dur = e.get("duration", 1.0)
                if onset > cursor:
                    lines.append(_rest_xml(onset - cursor))
                    cursor = onset
                pitch = e.get("pitch", 60)
                syll = e.get("syllable")
                melisma = e.get("melisma", False)
                syllabic_val = None
                if syll:
                    syllabic_val = "begin" if melisma else "single"
                elif melisma:
                    syllabic_val = "middle"
                lines.append(_note_xml(pitch, dur, syll, syllabic_val))
                cursor = onset + dur

            if cursor < 4.0:
                lines.append(_rest_xml(4.0 - cursor))

            lines.append('    </measure>')

        lines.extend(['  </part>', '</score-partwise>'])
        return "\n".join(lines)


def export_compiled_song_to_musicxml(compiled_song) -> str:
    """
    Export CompiledSong to full MusicXML.
    Includes work-title, part-list, measures, melody notes, lyrics, harmony.
    """
    try:
        from .musicxml_contracts import DEFAULT_DIVISIONS
    except ImportError:
        from musicxml_contracts import DEFAULT_DIVISIONS

    constraints = compiled_song.metadata.get("musicxml_constraints") if compiled_song.metadata else None
    hints = compiled_song.metadata.get("export_hints") if compiled_song.metadata else None
    divisions = getattr(constraints, "divisions", None) or DEFAULT_DIVISIONS
    key_center = getattr(hints, "key_center", "C") if hints else "C"
    tempo = getattr(hints, "tempo", 90) if hints else 90

    key_fifths = {"C": 0, "G": 1, "D": 2, "A": 3, "E": 4, "B": 5, "F": -1, "Bb": -2, "Eb": -3, "Ab": -4, "Db": -5,
                  "Am": 0, "Em": 1, "Bm": 2, "F#m": 3, "C#m": 4, "G#m": 5, "Dm": -1, "Gm": -2, "Cm": -3, "Fm": -4, "Bbm": -5}
    fifths = key_fifths.get(key_center, 0)

    by_measure: Dict[int, List[Dict]] = {}
    for sec in compiled_song.sections:
        for e in sec.melody_events:
            m = e.get("measure", 0)
            if m not in by_measure:
                by_measure[m] = []
            ev = dict(e)
            ev["_section_id"] = sec.section_id
            ev["_lyric_lines"] = sec.lyric_lines
            by_measure[m].append(ev)

    chords_by_measure: Dict[int, List[Dict]] = {}
    for sec in compiled_song.sections:
        for h in sec.harmony:
            m = h.get("measure", 0)
            if m not in chords_by_measure:
                chords_by_measure[m] = []
            chords_by_measure[m].append(h)

    total_bars = max(by_measure.keys(), default=0) + 1
    total_bars = max(total_bars, max(chords_by_measure.keys(), default=0) + 1)

    def _get_syllables(lyric_lines: List[str]) -> List[str]:
        words = []
        for line in lyric_lines or []:
            words.extend(str(line).split())
        return words

    syl_index_by_section: Dict[str, int] = {}
    syls_by_section: Dict[str, List[str]] = {}
    for sec in compiled_song.sections:
        syls_by_section[sec.section_id] = _get_syllables(sec.lyric_lines)
        syl_index_by_section[sec.section_id] = 0

    for m in range(total_bars):
        evs = sorted(by_measure.get(m, []), key=lambda x: (x.get("beat_position", 0), x.get("id", "")))
        for ev in evs:
            sid = ev.get("_section_id", "")
            if sid and syl_index_by_section.get(sid, 0) < len(syls_by_section.get(sid, [])):
                ev["_syllable"] = syls_by_section[sid][syl_index_by_section[sid]]
                syl_index_by_section[sid] += 1

    lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        '<score-partwise version="4.0">',
        '  <work><work-title>' + _escape(compiled_song.title) + '</work-title></work>',
        '  <identification><encoding><software>Song IR Engine</software></encoding></identification>',
        '  <part-list>',
        '    <score-part id="P1"><part-name>Voice</part-name></score-part>',
        '  </part-list>',
        '  <part id="P1">',
    ]

    for m in range(total_bars):
        lines.append(f'    <measure number="{m + 1}">')
        if m == 0:
            lines.append(f'      <attributes><divisions>{divisions}</divisions>')
            lines.append(f'        <key><fifths>{fifths}</fifths><mode>major</mode></key>')
            lines.append('        <time><beats>4</beats><beat-type>4</beat-type></time>')
            lines.append('        <clef><sign>G</sign><line>2</line></clef></attributes>')
            lines.append(f'      <sound tempo="{tempo}"/>')

        evs = sorted(by_measure.get(m, []), key=lambda x: (x.get("beat_position", 0), x.get("id", "")))

        for h in chords_by_measure.get(m, []):
            sym = h.get("symbol", "C") or "C"
            root_step = sym[0] if sym and sym[0].isalpha() else "C"
            alter = 1 if len(sym) > 1 and sym[1] == "#" else -1 if len(sym) > 1 and sym[1] == "b" else 0
            alter_tag = f"<alter>{alter}</alter>" if alter != 0 else ""
            kind = "minor" if "m" in sym and "maj" not in sym else "major"
            lines.append(f'      <harmony><root><root-step>{root_step}</root-step>{alter_tag}</root><kind>{kind}</kind></harmony>')

        cursor = 0.0
        for ev in evs:
            onset = ev.get("beat_position", 0)
            dur = ev.get("duration", 1.0)
            if onset > cursor:
                divs = max(1, int((onset - cursor) * divisions / 4))
                lines.append(f'      <note><rest/><duration>{divs}</duration><type>quarter</type></note>')
                cursor = onset
            pitch = ev.get("pitch", 60)
            step, alter, octave = _spell_pitch(pitch)
            alter_tag = f"<alter>{alter}</alter>" if alter != 0 else ""
            divs = max(1, int(dur * divisions / 4))
            typ = "quarter" if divs <= 4 else "half" if divs <= 8 else "whole"
            syll = ev.get("_syllable")
            lyric = f'<lyric><syllabic>single</syllabic><text>{_escape(syll)}</text></lyric>' if syll else ""
            lines.append(
                f'      <note><pitch><step>{step}</step>{alter_tag}<octave>{octave}</octave></pitch>'
                f'<duration>{divs}</duration><type>{typ}</type>{lyric}</note>'
            )
            cursor = onset + dur

        if cursor < 4.0:
            divs = max(1, int((4.0 - cursor) * divisions / 4))
            lines.append(f'      <note><rest/><duration>{divs}</duration><type>quarter</type></note>')

        lines.append('    </measure>')

    lines.extend(['  </part>', '</score-partwise>'])
    return "\n".join(lines)


def _escape(s: str) -> str:
    """Escape XML special chars."""
    if not s:
        return ""
    return str(s).replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace('"', "&quot;")

File: engines/composer-os-v2/core/export/musicxmlExporter.ts
/**
 * Composer OS V2 — MusicXML exporter
 * Converts ScoreModel to MusicXML. Single source of truth is the score model.
 * Durations: fixed divisions/quarter (480); <duration> and <type>/<dot/> from one tick value (musicXmlTickEncoding).
 */

import * as fs from 'fs';

import type { MusicXmlExportResult, MusicXmlExportOptions } from './exportTypes';
import { normalizeChordToken } from '../harmony/chordProgressionParser';
import type { ScoreModel, PartModel, MeasureModel } from '../score-model/scoreModelTypes';
import { DIVISIONS, MEASURE_DIVISIONS } from '../score-model/scoreModelTypes';
import { musicXmlKindContentFromKindText, parseChordForMusicXmlHarmony } from './chordSymbolMusicXml';
import {
  GUITAR_BASS_DUO_BASS_INSTRUMENT_SOUND,
  GUITAR_BASS_DUO_BASS_PART_NAME,
} from '../instrument-profiles/guitarBassDuoExportNames';
import {
  beatSpanToTicks,
  decomposeTicksToMinimalParts,
  decomposeTicksToStandardParts,
  typeAndDotsXml,
  tickSpecForLength,
} from './musicXmlTickEncoding';

function partDisplayNameForExport(p: PartModel): string {
  if (p.instrumentIdentity === 'acoustic_upright_bass') {
    return GUITAR_BASS_DUO_BASS_PART_NAME;
  }
  return p.name;
}

function instrumentSoundXmlForPart(p: PartModel): string {
  if (p.instrumentIdentity === 'acoustic_upright_bass') {
    return `<instrument-sound>${GUITAR_BASS_DUO_BASS_INSTRUMENT_SOUND}</instrument-sound>`;
  }
  return '';
}

function escapeXml(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/'/g, '&apos;');
}

/** V3.4e — stable bar order for export; MusicXML `number` is assigned 1..n separately (Sibelius-safe). */
function measuresInExportOrder(part: PartModel): MeasureModel[] {
  return [...part.measures].sort((a, b) => a.index - b.index);
}

function midiToPitch(midi: number): { step: string; alter: number; octave: number } {
  const semitones = ((midi % 12) + 12) % 12;
  const octave = Math.floor(midi / 12) - 1;
  const stepMap: Record<number, { step: string; alter: number }> = {
    0: { step: 'C', alter: 0 }, 1: { step: 'C', alter: 1 }, 2: { step: 'D', alter: 0 },
    3: { step: 'D', alter: 1 }, 4: { step: 'E', alter: 0 }, 5: { step: 'F', alter: 0 },
    6: { step: 'F', alter: 1 }, 7: { step: 'G', alter: 0 }, 8: { step: 'G', alter: 1 },
    9: { step: 'A', alter: 0 }, 10: { step: 'A', alter: 1 }, 11: { step: 'B', alter: 0 },
  };
  const { step, alter } = stepMap[semitones];
  return { step, alter, octave };
}

function tieXmlForSplitIndex(i: number, n: number): string {
  if (n <= 1) return '';
  if (i === 0) return '<tie type="start"/>';
  if (i === n - 1) return '<tie type="stop"/>';
  return '<tie type="stop"/><tie type="start"/>';
}

function emitRestTicks(parts: number[], voice: number): string {
  let xml = '';
  for (const t of parts) {
    const spec = tickSpecForLength(t);
    xml += `        <note><rest/><duration>${t}</duration>${typeAndDotsXml(spec)}<voice>${voice}</voice></note>\n`;
  }
  return xml;
}

function emitPitchedTicks(
  parts: number[],
  voice: number,
  pitchXml: string,
  articulationXml: string,
  dynAttr: string
): string {
  let xml = '';
  const n = parts.length;
  for (let i = 0; i < n; i++) {
    const t = parts[i];
    const spec = tickSpecForLength(t);
    const tie = tieXmlForSplitIndex(i, n);
    xml += `        <note${dynAttr}>${pitchXml}<duration>${t}</duration>${tie}${typeAndDotsXml(spec)}<voice>${voice}</voice>${i === 0 ? articulationXml : ''}</note>\n`;
  }
  return xml;
}

/**
 * Same tick arithmetic as `eventsToXml` for one voice — parity with written XML.
 */
export function computeVoiceExportDivisionSum(measure: MeasureModel, voice: number): number {
  const sorted = [...measure.events]
    .filter((e) => e.kind === 'note' || e.kind === 'rest')
    .sort((a, b) => a.startBeat - b.startBeat);
  const voiceEvents = sorted.filter((e) => (e.voice ?? 1) === voice);
  let cursor = 0;
  let sum = 0;
  for (const e of voiceEvents) {
    const sb = e.startBeat;
    const eb = sb + e.duration;
    const { start: startDiv, end: endDiv } = beatSpanToTicks(sb, eb);
    const safeStart = Math.max(startDiv, cursor);
    if (endDiv <= safeStart) continue;
    if (safeStart > cursor) {
      const restDiv = safeStart - cursor;
      sum += restDiv;
      cursor = safeStart;
    }
    const durDiv = endDiv - safeStart;
    sum += durDiv;
    cursor = endDiv;
  }
  if (cursor < MEASURE_DIVISIONS) {
    sum += MEASURE_DIVISIONS - cursor;
  }
  return sum;
}

function eventsToXml(
  measure: MeasureModel,
  measureIndex: number,
  part: PartModel,
  decompose: (ticks: number) => number[],
  opts: MusicXmlExportOptions
): string {
  const v2count = measure.events.filter((e) => (e as any).voice === 2).length;
  if (v2count > 0) {
    const logLine = `[pre-export] measure=${measureIndex+1} voice2events=${v2count}\n`;
    try {
      fs.appendFileSync('C:/Users/mike/composer-debug.log', logLine);
    } catch {}
  }
  const partId = part.id;
  let sorted = [...measure.events]
    .filter((e) => e.kind === 'note' || e.kind === 'rest')
    .sort((a, b) => a.startBeat - b.startBeat);
  if (opts.bassStaffVoice1Only && part.instrumentIdentity === 'acoustic_upright_bass') {
    sorted = sorted.map((e) => ({ ...e, voice: 1 }));
  }
  const voices = [...new Set(sorted.map((e) => e.voice ?? 1))].sort((a, b) => a - b);
  const debugTicks = typeof process !== 'undefined' && process.env?.COMPOSER_OS_DEBUG_MUSICXML_TICKS === '1';

  let xml = '';
  for (let vi = 0; vi < voices.length; vi++) {
    const voice = voices[vi];
    if (vi > 0) {
      xml += `        <backup><duration>${MEASURE_DIVISIONS}</duration></backup>\n`;
    }
    let cursor = 0;
    cursor = 0;
    const voiceEvents = sorted.filter((e) => (e.voice ?? 1) === voice);
    let voiceTickSum = 0;

    for (const e of voiceEvents) {
      const sb = e.startBeat;
      const eb = sb + e.duration;
      const { start: startDiv, end: endDiv } = beatSpanToTicks(sb, eb);
      const safeStart = Math.max(startDiv, cursor);
      if (endDiv <= safeStart) {
        continue;
      }

      if (safeStart > cursor) {
        const restDiv = safeStart - cursor;
        const parts = decompose(restDiv);
        xml += emitRestTicks(parts, voice);
        for (const p of parts) voiceTickSum += p;
        cursor = safeStart;
      }

      const durDiv = endDiv - safeStart;
      const parts = decompose(durDiv);
      if (e.kind === 'rest') {
        xml += emitRestTicks(parts, voice);
      } else {
        const { step, alter, octave } = midiToPitch(e.pitch);
        const alterEl = alter !== 0 ? `<alter>${alter}</alter>` : '';
        const pitchXml = `<pitch><step>${step}</step>${alterEl}<octave>${octave}</octave></pitch>`;
        const art = (e as { articulation?: string }).articulation;
        const notationsEl = art ? `<notations><articulations><${art}/></articulations></notations>` : '';
        const vel = (e as { velocity?: number }).velocity;
        const dynAttr =
          vel !== undefined
            ? ` dynamics="${Math.min(100, Math.max(0, (vel / 127) * 100)).toFixed(2)}"`
            : '';
        xml += emitPitchedTicks(parts, voice, pitchXml, notationsEl, dynAttr);
      }
      for (const p of parts) voiceTickSum += p;
      cursor = endDiv;
    }

    if (cursor < MEASURE_DIVISIONS) {
      const restDiv = MEASURE_DIVISIONS - cursor;
      const parts = decompose(restDiv);
      xml += emitRestTicks(parts, voice);
      for (const p of parts) voiceTickSum += p;
    }

    if (voiceTickSum !== MEASURE_DIVISIONS) {
      if (voice !== 1) {
        console.warn(`[wyble-export] voice ${voice} tick sum ${voiceTickSum} !== ${MEASURE_DIVISIONS} in measure ${measureIndex + 1} — skipping voice`);
        continue;
      }
      throw new Error(
        `MusicXML export: part ${partId} measure ${measureIndex + 1} voice ${voice}: tick sum ${voiceTickSum} ≠ expected ${MEASURE_DIVISIONS} (divisions×beats=${DIVISIONS}×4)`
      );
    }
    if (debugTicks) {
      console.log(
        `[MusicXML ticks] part=${partId} measure=${measureIndex + 1} voice=${voice} sum=${voiceTickSum} expected=${MEASURE_DIVISIONS}`
      );
    }
  }

  return xml;
}

/** Export ScoreModel to MusicXML string. */
export function exportScoreModelToMusicXml(score: ScoreModel, options?: MusicXmlExportOptions): MusicXmlExportResult {
  try {
    const opts: MusicXmlExportOptions = options ?? {};
    const decompose = opts.minimizeNoteFragmentation ? decomposeTicksToMinimalParts : decomposeTicksToStandardParts;
    const dbg = score.keySignatureExportDebug;
    const ks0 = score.keySignature;
    if (typeof process !== 'undefined' && process.env?.COMPOSER_OS_DEBUG_KEY_XML === '1') {
      console.log('[V3.4c exportScoreModelToMusicXml]', {
        inferredKey: dbg?.inferredKey,
        inferredFifths: dbg?.inferredFifths,
        exportKeyWritten: dbg?.exportKeyWritten,
        xmlFifths: ks0?.fifths ?? 0,
        xmlMode: ks0?.mode ?? 'major',
        xmlHideKey: ks0?.hideKeySignature ?? false,
      });
    }

    const partList = score.parts
      .map((p) => {
        const mxProg = Math.min(128, Math.max(1, p.midiProgram + 1));
        const display = partDisplayNameForExport(p);
        const soundEl = instrumentSoundXmlForPart(p);
        return `    <score-part id="${p.id}">
      <part-name>${escapeXml(display)}</part-name>
      <score-instrument id="${p.id}-I1"><instrument-name>${escapeXml(display)}</instrument-name>${soundEl ? `\n        ${soundEl}` : ''}</score-instrument>
      <midi-instrument id="${p.id}-I1">
        <midi-program>${mxProg}</midi-program>
      </midi-instrument>
    </score-part>`;
      })
      .join('\n');

    let xml = `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 3.1 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise version="3.1">
  <work><work-title>${escapeXml(score.title)}</work-title></work>
  <part-list>
${partList}
  </part-list>
`;

    for (let partIndex = 0; partIndex < score.parts.length; partIndex++) {
      const part = score.parts[partIndex];
      const clefSign = part.clef === 'bass' ? 'F' : 'G';
      const clefLine = part.clef === 'bass' ? 4 : 2;
      xml += `  <part id="${part.id}">\n`;

      const orderedMeasures = measuresInExportOrder(part);
      for (let i = 0; i < orderedMeasures.length; i++) {
        const m = orderedMeasures[i];
        const measureNumber = i + 1;
        xml += `  <measure number="${measureNumber}">\n`;

        if (i === 0) {
          const tempoEl = score.tempo ? `\n    <sound tempo="${score.tempo}"/>` : '';
          const feelEl =
            partIndex === 0 && score.feelProfile
              ? `    <direction placement="below"><direction-type><words>${escapeXml(
                  `Feel: ${score.feelProfile.tempoFeel}; straight notation (export); duo`
                )}</words></direction-type></direction>\n`
              : '';
          const ks = score.keySignature;
          const fifths = ks?.fifths ?? 0;
          const keyMode = ks?.mode ?? 'major';
          const hideKey = ks?.hideKeySignature ?? false;
          const printAttr = hideKey ? ' print-object="no"' : '';
          const keyCaption =
            partIndex === 0 && ks?.caption
              ? `    <direction placement="above"><direction-type><words>${escapeXml(ks.caption)}</words></direction-type></direction>\n`
              : '';
          xml += `    <attributes>
      <divisions>${DIVISIONS}</divisions>
      <key${printAttr}><fifths>${fifths}</fifths><mode>${keyMode}</mode></key>
      <time><beats>4</beats><beat-type>4</beat-type></time>
      <clef><sign>${clefSign}</sign><line>${clefLine}</line></clef>
    </attributes>${tempoEl}
${keyCaption}${feelEl}`;
        }

        if (m.rehearsalMark) {
          xml += `    <direction placement="above"><direction-type><rehearsal>${escapeXml(m.rehearsalMark)}</rehearsal></direction-type></direction>\n`;
        }
        if (m.chord && !opts.omitChordSymbols) {
          const assertLocked = opts.assertLockedHarmonyBars;
          if (assertLocked && assertLocked.length === orderedMeasures.length && i < assertLocked.length) {
            const exp = assertLocked[i];
            if (normalizeChordToken(m.chord) !== normalizeChordToken(exp ?? '')) {
              throw new Error(
                `CUSTOM HARMONY NOT REACHING GOLDEN PATH: MusicXML export bar ${measureNumber} chord "${m.chord}" !== locked "${exp}"`
              );
            }
          }
          const { rootStep, rootAlter, kindText, bassStep, bassAlter } = parseChordForMusicXmlHarmony(m.chord, {
            literalKind: opts.preserveChordKindLiterals === true,
          });
          const kindContent = musicXmlKindContentFromKindText(kindText);
          const alterEl = rootAlter !== 0 ? `<root-alter>${rootAlter}</root-alter>` : '';
          const bassAlterEl =
            bassStep !== undefined && bassAlter !== undefined && bassAlter !== 0
              ? `<bass-alter>${bassAlter}</bass-alter>`
              : '';
          const bassEl =
            bassStep !== undefined
              ? `<bass><bass-step>${bassStep}</bass-step>${bassAlterEl}</bass>`
              : '';
          xml += `    <harmony><root><root-step>${rootStep}</root-step>${alterEl}</root><kind text="${escapeXml(kindText)}">${escapeXml(kindContent)}</kind>${bassEl}</harmony>\n`;
        }

        xml += eventsToXml(m, i, part, decompose, opts);
        xml += `  </measure>\n`;
      }
      xml += `  </part>\n`;
    }

    xml += `</score-partwise>
`;
    return { success: true, xml, errors: [] };
  } catch (e) {
    return {
      success: false,
      errors: [e instanceof Error ? e.message : String(e)],
    };
  }
}

/** Legacy stub — delegates to exportScoreModelToMusicXml when input is ScoreModel. */
export function exportToMusicXml(input: unknown, options?: MusicXmlExportOptions): MusicXmlExportResult {
  if (input && typeof input === 'object' && 'title' in input && 'parts' in input) {
    return exportScoreModelToMusicXml(input as ScoreModel, options);
  }
  return {
    success: false,
    errors: ['Input must be ScoreModel'],
  };
}
